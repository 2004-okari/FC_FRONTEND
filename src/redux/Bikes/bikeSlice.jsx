import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const BASE_URL = 'http://127.0.0.1:3000/';

const tokenSelector = (state) => state.auth.token;

export const getBikes = createAsyncThunk(
  'bikes/getBikes',
  async (_, thunkAPI) => {
    try {
      const token = tokenSelector(thunkAPI.getState());
      const resp = await axios.get(`${BASE_URL}display_bikes`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      return resp.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error);
    }
  },
);

export const createBike = createAsyncThunk(
  'bike/createBike',
  async (bike, thunkAPI) => {
    try {
      const token = tokenSelector(thunkAPI.getState());
      const response = await axios.post(`${BASE_URL}create_bikes`, { bike }, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const deleteBike = createAsyncThunk(
  'bike/deleteBike',
  async (bikeId, thunkAPI) => {
    try {
      const token = tokenSelector(thunkAPI.getState());
      await axios.delete(`${BASE_URL}delete_bike/${bikeId}`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      return bikeId;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

const initialState = {
  message: '',
  isLoading: false,
  error: undefined,
};

const bikeSlice = createSlice({
  name: 'bikes',
  initialState,
  reducers: {},
  extraReducers(builder) {
    builder
      .addCase(getBikes.pending, (state) => ({ ...state, isLoading: true }))
      .addCase(getBikes.fulfilled, (state, action) => ({
        ...state,
        isLoading: false,
        message: action.payload,
      }))
      .addCase(getBikes.rejected, (state, action) => ({
        ...state,
        isLoading: false,
        error: action.payload.message
          ? action.payload.message
          : 'An error occurred',
      }))

      .addCase(createBike.pending, (state) => ({ ...state, isLoading: true }))
      .addCase(createBike.fulfilled, (state, action) => ({
        ...state,
        isLoading: false,
        message: action.payload,
      }))
      .addCase(createBike.rejected, (state, action) => ({
        ...state,
        isLoading: false,
        error: action.payload.message
          ? action.payload.message
          : 'An error occurred',
      }))

      .addCase(deleteBike.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(deleteBike.fulfilled, (state, action) => {
        const updatedBikes = state.message.bikes.filter(
          (bike) => bike.id !== action.payload,
        );
        return {
          ...state,
          isLoading: false,
          message: {
            ...state.message,
            bikes: updatedBikes,
          },
        };
      })
      .addCase(deleteBike.rejected, (state, action) => ({
        ...state,
        isLoading: false,
        error: action.payload.message
          ? action.payload.message
          : 'An error occurred',
      }));
  },
});

export default bikeSlice.reducer;
