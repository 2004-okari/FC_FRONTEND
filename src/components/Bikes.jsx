import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { getBikes } from '../redux/Bikes/bikeSlice';

const Bikes = () => {
  const dispatch = useDispatch();
  const { message, isLoading, error } = useSelector((store) => store.bikes);

  useEffect(() => {
    dispatch(getBikes());
  }, [dispatch]);

  const truncateText = (text, maxLines) => {
    if (!text) {
      return '';
    }
    const lines = text.split(' ').slice(0, maxLines);
    return lines.join('\n');
  };

  const leftClick = () => {
    const box = document.querySelector('.carosoul');
    box.scrollLeft -= 310;
  };

  const rightClick = () => {
    const box = document.querySelector('.carosoul');
    box.scrollLeft += 310;
  };

  if (isLoading) {
    return <div className="flex items-center justify-center w-full">Loading</div>;
  }

  if (error) {
    return (
      <p>
        Something went wrong!
        {error}
      </p>
    );
  }

  if (message) {
    return (
      <div className="relative flex flex-col overflow-hidden w-full items-center justify-center p-6 gap-8">
        <div className="p-4 carosoul flex overflow-x-hidden w-full scroll-smooth gap-4">
          {message.bikes.map((bike) => (
            <Link to={`/show/${bike.id}`} key={bike.id}>
              <div className="flex flex-col bg-[#F9F9F9] w-[300px] gap-1 border-solid border-2 p-1 rounded-lg border-[#E1E1E1]" key={bike.id}>
                <img src={bike.image} alt="" className="min-h-[300px] rounded-lg" />
                <p className="text-center text-xl font-bold">{bike.name}</p>
                <p className="text-lg">
                  {truncateText(bike.description, 20)}
                  <span className="text-blue-500 ml-1">Read More...</span>
                </p>
              </div>
            </Link>
          ))}
        </div>
        <div className="flex items-center gap-5">
          <button type="button" className="absolute top-[50%] left-1 p-2 border-2 bg-white font-bold rounded-full shadow-2xl" onClick={leftClick}>
            &lt;
          </button>
          <button type="button" className="absolute top-[50%] right-1 p-2 border-2 bg-white font-bold rounded-full shadow-2xl" onClick={rightClick}>
            &gt;
          </button>
        </div>
      </div>
    );
  }

  return <div>No Bikes to show</div>;
};

export default Bikes;