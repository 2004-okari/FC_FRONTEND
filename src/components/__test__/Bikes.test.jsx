import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import Bikes from "../Bikes";

const mockStore = configureStore([]);

describe("Cars Component", () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      bikes: {
        message: null,
        isLoading: false,
        error: null,
      },
    });

    store.dispatch = jest.fn();

    component = render(
      <Provider store={store}>
        <MemoryRouter>
          <Bikes />
        </MemoryRouter>
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(component.container).toMatchSnapshot();
  });
});
