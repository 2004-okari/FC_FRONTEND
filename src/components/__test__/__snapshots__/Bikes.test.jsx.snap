// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cars Component renders correctly 1`] = `
<div>
  <div>
    No Bikes to show
  </div>
</div>
`;
